encoding/dot/dot_test.go

```diff
@@ -0,0 +1,1436 @@
+			}
+			nv := namedAttrNode{id: v, name: alpha[v : v+1], attr: at}
+			dg.AddUndirectedEdge(concrete.Edge{F: nu, T: nv}, 0)
+		}
+	}
+	return dg
+}
+
+type attrEdge struct {
+	from, to graph.Node
+
+	attr []Attribute
+}
+
+func (e attrEdge) From() graph.Node           { return e.from }
+func (e attrEdge) To() graph.Node             { return e.to }
+func (e attrEdge) DOTAttributes() []Attribute { return e.attr }
+
+func directedEdgeAttrGraphFrom(g []set, attr map[edge][]Attribute) graph.DirectedGraph {
+	dg := concrete.NewDirectedGraph()
+	for u, e := range g {
+		for v := range e {
+			dg.AddDirectedEdge(attrEdge{from: concrete.Node(u), to: concrete.Node(v), attr: attr[edge{from: u, to: v}]}, 0)
+		}
+	}
+	return dg
+}
+
+func undirectedEdgeAttrGraphFrom(g []set, attr map[edge][]Attribute) graph.Graph {
+	dg := concrete.NewGraph()
+	for u, e := range g {
+		for v := range e {
+			dg.AddUndirectedEdge(attrEdge{from: concrete.Node(u), to: concrete.Node(v), attr: attr[edge{from: u, to: v}]}, 0)
+		}
+	}
+	return dg
+}
+
+type portedEdge struct {
+	from, to graph.Node
+
+	directed bool
+
+	fromPort    string
+	fromCompass string
+	toPort      string
+	toCompass   string
+}
+
+func (e portedEdge) From() graph.Node { return e.from }
+func (e portedEdge) To() graph.Node   { return e.to }
+
+// TODO(kortschak): Figure out a better way to handle the fact that
+// headedness is an undefined concept in undirected graphs. We sort
+// nodes by ID, so lower ID nodes are always from nodes in undirected
+// graphs. We can probably do this in the printer, but I am leaving
+// this here as a WARNING.
+// Maybe the approach should be to document that for undirected graphs
+// the low ID node should be returned by the FromPort and the high ID
+// by the ToPort calls.
+func (e portedEdge) FromPort() (port, compass string) {
+	return e.fromPort, e.fromCompass
+}
+func (e portedEdge) ToPort() (port, compass string) {
+	return e.toPort, e.toCompass
+}
+
+func directedPortedAttrGraphFrom(g []set, attr [][]Attribute, ports map[edge]portedEdge) graph.DirectedGraph {
+	dg := concrete.NewDirectedGraph()
+	for u, e := range g {
+		var at []Attribute
+		if u < len(attr) {
+			at = attr[u]
+		}
+		nu := attrNode{id: u, attr: at}
+		for v := range e {
+			if v < len(attr) {
+				at = attr[v]
+			}
+			pe := ports[edge{from: u, to: v}]
+			pe.from = nu
+			pe.to = attrNode{id: v, attr: at}
+			dg.AddDirectedEdge(pe, 0)
+		}
+	}
+	return dg
+}
+
+func undirectedPortedAttrGraphFrom(g []set, attr [][]Attribute, ports map[edge]portedEdge) graph.Graph {
+	dg := concrete.NewGraph()
+	for u, e := range g {
+		var at []Attribute
+		if u < len(attr) {
+			at = attr[u]
+		}
+		nu := attrNode{id: u, attr: at}
+		for v := range e {
+			if v < len(attr) {
+				at = attr[v]
+			}
+			pe := ports[edge{from: u, to: v}]
+			pe.from = nu
+			pe.to = attrNode{id: v, attr: at}
+			dg.AddUndirectedEdge(pe, 0)
+		}
+	}
+	return dg
+}
+
+type graphAttributer struct {
+	graph.Graph
+	gobal attributer
```

Comment:
Good catch. That should not be there.

For reference, all this comes from [this](http://www.graphviz.org/doc/info/lang.html) and playing around.